samples: drivers: added a simple touch controller sample

fkokosinski opened the pull request:
This PR adds a sample for the touch controller. The sample is supposed to help examine issues when working with touchscreens. It draws a plus in the last touched position.

Can be tested natively with:
```
west build -p -b native_posix/native/64 samples/drivers/touch -t run
```

laudominik committed:
samples: drivers: added a simple touch controller sample

The sample is supposed to help examine the issues with touchscreen.
It draws a plus in last touched position.

Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented:
Hey @kartben @faxe1008, thanks for your comments.

Ideally, we'd like this demo to work on smaller targets as well, where e.g. there's not enough space for LVGL. Plus, we think that having a kind of "minimal" demo with no additional library overhead can be beneficial when working on a new input driver, or debugging an old one. So we'd opt for keeping this demo separate from the one you've linked.

Let me know what you guys think. Thanks

fkokosinski commented:
Hey @kartben, let me know if there's something blocking us from moving forward with this PR. Thanks!

gmarull commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+#define BUFFER_SIZE  ((CROSS_WIDTH) * (CROSS_HEIGHT) * 4)
+#define REFRESH_RATE 100
+
+#ifndef DT_CHOSEN_zephyr_touch
+#error "Unsupported board: zephyr,touch is not assigned"
+#endif
+
+#ifndef DT_CHOSEN_zephyr_display
+#error "Unsupported board: zephyr,display is not assigned"
+#endif
```
DT raw macros can't be used directly

fkokosinski replied:
Switched to using `DT_NODE_EXISTS`

gmarull commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sample, LOG_LEVEL_INF);
```
please, move after includes..

fkokosinski replied:
Done

gmarull commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+
+	if (bpp == 0 || bpp > 4) {
+		LOG_ERR("Unsupported BPP=%d", bpp);
+		return 1;
```
main can only return 0 (unless this changed)

fkokosinski replied:
Fixed

gmarull commented on samples/drivers/touch/README.rst:
```diff
@@ -0,0 +1,32 @@
+.. zephyr:code-sample:: touch
+   :name: touchs
+   :relevant-api: io_interfaces
```
this is not an API

faxe1008 commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+	display_blanking_off(display_dev);
+
+	clear_screen();
+	tick_count = 0;
```
This is never read from

fkokosinski replied:
Removed

faxe1008 commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+int main(void)
+{
+	size_t tick_count;
+	size_t bpp;
```
If the api does not require it, I would avoid using posix types.

fkokosinski replied:
Changed

fabiobaltieri commented on samples/drivers/touch/boards/native_posix_64.overlay:
```diff
@@ -0,0 +1,6 @@
+/ {
+	chosen {
+		zephyr,display = &sdl_dc;
+		zephyr,touch = &input_sdl_touch;
```
the first is already in `boards/native/native_sim/native_sim.dts`, it should not be needed here, the second should probably go in the board file, @kartben what do you think?

fabiobaltieri commented on samples/drivers/touch/boards/stm32f429i_disc1.overlay:
```diff
@@ -0,0 +1,5 @@
+/ {
+	chosen {
+		zephyr,touch = &stmpe811;
```
same, I think it's fair to add this in the board file(s)

fabiobaltieri commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+
+#define CROSS_WIDTH  20
+#define CROSS_HEIGHT 20
+#define BUFFER_SIZE  ((CROSS_WIDTH) * (CROSS_HEIGHT) * 4)
```
```suggestion
#define BUFFER_SIZE  (CROSS_WIDTH * CROSS_HEIGHT * 4)
```

the convention is to parenthesize the definition, no need to parenthesize everything

fkokosinski replied:
Fixed

fabiobaltieri commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+						    .width = CROSS_WIDTH,
+						    .height = CROSS_HEIGHT};
+
+static uint8_t buffer_cross[BUFFER_SIZE] = {0};
+static const uint8_t buffer_cross_empty[BUFFER_SIZE] = {0};
```
drop the initializers, these are already zero initialized automatically

fkokosinski replied:
Zero initializers were dropped

fabiobaltieri commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+
+static void touch_event_callback(struct input_event *evt, void *user_data)
+{
+	if (evt->code == INPUT_ABS_X) {
+		touch_point.x = evt->value;
+		k_sem_give(&sync_x);
+	}
+	if (evt->code == INPUT_ABS_Y) {
+		touch_point.y = evt->value;
+		k_sem_give(&sync_y);
+	}
```
that's not quite the right way of using the API, you should cache x y locally and generate the event once you see an event with the sync bit set, somewhat like this https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/kscan/kscan_input.c#L26-L50, then you only need some semaphore

fkokosinski replied:
Thanks for the tip, should be fixed now

fkokosinski commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+static struct {
+	size_t x;
+	size_t y;
+} touch_point, touch_point_drawn;
```
Hi, we need to keep track of where the last glyph was placed, so that we can remove it when we draw the next one, hence the two structures.

fkokosinski commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+	while (1) {
+		k_msleep(REFRESH_RATE);
+		k_sem_take(&sync_x, K_FOREVER);
+		k_sem_take(&sync_y, K_FOREVER);
```
Fixed

fkokosinski commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+
+	k_sem_init(&sync_x, 0, 1);
+	k_sem_init(&sync_y, 0, 1);
+	INPUT_CALLBACK_DEFINE(NULL, touch_event_callback, NULL);
```
Moved outside of function

fkokosinski commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,142 @@
+#include <zephyr/drivers/display.h>
+
+#define CROSS_WIDTH  20
+#define CROSS_HEIGHT 20
```
Moved to Kconfig

fkokosinski commented on samples/drivers/touch/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+  name: touch_sample
+tests:
+  sample.touch.native:
+    platform_allow: native_posix/native/64
```
Moved to `native_sim`

fkokosinski commented:
Hi @gmarull @faxe1008, I believe your comments were addressed. Could you PTAL? Thanks!

@kartben could you take a look at this PR? If I'm not mistaken, your +1 is needed, as you're the assignee here.

kartben commented on samples/drivers/touch/README.rst:
```diff
@@ -0,0 +1,31 @@
+.. zephyr:code-sample:: touch
+   :name: touch
+
+   Basic touchscreen sample.
```
This needs a proper name and description should be worded as if completing the sentence "This code sample shows how to ..."

```suggestion
.. zephyr:code-sample:: draw_touch_events
   :name: Draw touch events
   :relevant-api: input_events display_interface

   Visualize touch events on a display.
```

fkokosinski replied:
Done.

kartben commented on samples/drivers/touch/Kconfig:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2024 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config CROSS_WIDTH
+	int "Horizontal length of the cross to be drawn"
+	default 20
+
+config CROSS_HEIGHT
+	int "Vertical length of the cross to be drawn"
+	default 20
```
A nit but I think it would maybe make more sense to express size in terms of a ratio, this way there would be less chances people ever need to override the default value should their screen be really large/small.

fkokosinski replied:
Switched to ratio

kartben commented on samples/drivers/touch/README.rst:
```diff
@@ -0,0 +1,31 @@
+Below is an example on how to build the sample for :ref:`stm32f746g_disco_board`;
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/drivers/touch
```
please move the sample to `samples/subsys/input/draw_touch_events` (happy to hear about other suggestions, but finding a better name than the generic "touch" is important, and so is moving to a either input or display subsys)

fkokosinski replied:
Moved

kartben commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,137 @@
+
+LOG_MODULE_REGISTER(sample, LOG_LEVEL_INF);
+
+#define BUFFER_SIZE  (CONFIG_CROSS_WIDTH * CONFIG_CROSS_HEIGHT * 4)
```
As this can amount to a non-negligible amount of memory (1.6 KB) even on RGB565, monochrome, etc displays, it might be better to allocate based on bpp?

fkokosinski replied:
Switched to using BPP here

kartben reviewed (changes requested):
apologies for the delay in reviewing. Thanks for the contribution!

laudominik committed:
boards: added touch controller to the `/chosen` node

This adds zephyr,touch property to boards with touch controllers, analogous
to `zephyr,display`.

Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

faxe1008 commented on boards/st/stm32f746g_disco/stm32f746g_disco.dts:
```diff
@@ -24,6 +24,7 @@
 		zephyr,dtcm = &dtcm;
 		zephyr,flash-controller = &n25q128a1;
 		zephyr,display = &ltdc;
+		zephyr,touch = &ft5336;
```
I like this idea very much, this could be way forward to remove the lvgl-pointer-input from the intree board definitions. Initially this was done for ease of use coming from the old kscan-input for LVGL. 

I imagine adding a flag `CONFIG_LV_Z_POINTER_FROM_CHOSEN` which defaults to true and depends on LVGL. As a whole this will still get users a "just works" experience with LVGL while removing module dependencies from intree boards. Thanks for the impulse @kartben

fkokosinski replied:
I believe this change was applied.

faxe1008 reviewed (dismissed):
Looks very good. Only thing from my side: iirc there is a doc file which explains the different chosen nodes used in zephyr upstream. An entry there would be also good.

laudominik committed:
boards: stm32f429i_disc1: inverted touch controller axes

This inverts x and y axes as reported by stmpe811 driver for stm32f429i.

Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented on samples/drivers/touch/README.rst:
```diff
@@ -0,0 +1,31 @@
+Building and Running
+********************
+While this is a generic sample and it should work with any boards with both display controllers
+and touch controllers supported by Zephyr (provided the corresponding /chosen node properties
```
Addressed.

fkokosinski commented on samples/drivers/touch/src/main.c:
```diff
@@ -0,0 +1,137 @@
+	LOG_INF("Touch sample for touchscreen: %s, dc: %s", touch_dev->name, display_dev->name);
+
+	if (!device_is_ready(touch_dev)) {
+		LOG_ERR("Device %s not found. Aborting sample.", display_dev->name);
```
Addressed.

fkokosinski commented on samples/drivers/touch/README.rst:
```diff
@@ -0,0 +1,31 @@
+********************
+While this is a generic sample and it should work with any boards with both display controllers
+and touch controllers supported by Zephyr (provided the corresponding /chosen node properties
+are set i.e. `zephyr,touch` and `zephyr,display`).
```
Addressed.

kartben reviewed (approved):
Great work, thanks!
@fkokosinski in the future please refrain from marking comments as resolved yourself - it's rather inconvenient to have to go through all closed conversations to check and, more importantly, it's up to the original reviewer to confirm their recommendations have been properly addressed. https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers
Thanks!

fkokosinski commented:
> Great work, thanks! @fkokosinski in the future please refrain from marking comments as resolved yourself - it's rather inconvenient to have to go through all closed conversations to check and, more importantly, it's up to the original reviewer to confirm their recommendations have been properly addressed. https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers Thanks!

Right, sorry about that.  I'll keep this in mind in the future :)

fkokosinski commented:
The CI failure seems to be unrelated. On 8fda052826de630fc8a973325675759234a01fef (main) I seem to be getting the exact same error when running `west build -p -b esp32s3_touch_lcd_1_28/esp32s3/appcpu tests/lib/c_lib/common -T libraries.libc.common`.

fkokosinski commented:
Rebased due to failing CI.
